skrub/_table_vectorizer.py

```diff
@@ -495,147 +449,95 @@ def _clone_transformers(self):
             self.n_jobs,
         )
 
-    def _auto_cast(self, X: pd.DataFrame) -> pd.DataFrame:
-        """Takes a dataframe and tries to convert its columns to their best possible
-        data type.
+    def _auto_cast(self, X, reset=True):
+        """Convert each column of a dataframe to the "optimal" dtype \
+            for scikit-learn estimators.
+
+        The main heuristics are the following:
+        - pandas extension dtypes conversion to numpy dtype
+        - datetime conversion using ``skrub.to_datetime``
+        - numeric conversion using ``pandas.to_numeric``
+        - numeric columns with missing values are converted to float to input np.nan
+        - categorical columns dtypes are updated with the new entries (if any)
+          during transform.
 
         Parameters
         ----------
         X : :obj:`~pandas.DataFrame` of shape (n_samples, n_features)
             The data to be transformed.
 
+        reset : bool, default=True
+            If set to ``True`` (during fit), creates ``inferred_column_types_``,
+            the mapping between columns of the training dataframe and their types.
+            If set to ``False`` (during transform), updates ``inferred_column_types_``
+            for the categorical columns with the new categories seen during transform.
+
         Returns
         -------
-        :obj:`~pandas.DataFrame`
-            The same :obj:`~pandas.DataFrame`, with its columns cast to their
-            best possible data type.
+        X : :obj:`~pandas.DataFrame`
```

Comment:
Regarding that, if we want to avoid using numpy array in the future, we should avoid copying the input with scikit-learn check_array. One option would be to copy the dataframe in `auto_cast` and return a copy. IMHO, parsers shouldn't modify their inputs in place.
Let's discuss this in #848.